### PR TITLE
Side dist

### PR DESCRIPTION
**Pull request by jim-v**
Implements a side following mode.
Fixes some wierdness when operating in the map frame

### CODE REVIEW

**Comment by jim-v** [comment]
Run with UbiquityRobotics/magni_robot#86

File: src/map.cpp [hunk 1]
```diff
@@ -0,0 +1,936 @@
+/*
```
Comment (rohbotics) [inline src/map.cpp]: I think this is a mistake.

Unless there has been a major architecture change :)

File: README.md [hunk 2]
```diff
@@ -25,3 +25,30 @@ To run, give the following command:
 ## Node details
 
 Please refer to [the move_basic wiki page](http://wiki.ros.org/move_basic) for node documentation.
+
+## New obstacle avoidance behavoir (March 2019)
```
Comment (rohbotics) [inline README.md:29]: Can you describe the new behavior here

File: include/move_basic/obstacle_detector.h [hunk 3]
```diff
@@ -122,13 +122,19 @@ class ObstacleDetector
...
+   float obstacle_dist(bool forward, float &left_dist, float &right_dist,
+                       tf2::Vector3 &fl, tf2::Vector3 &fr);
 
    // return distance in radians to closest obstacle
    float obstacle_angle(bool left);
 
    // for testing
    void add_test_point(tf2::Vector3 p);
+
+   //TODO: refactor
```
Comment (rohbotics) [inline include/move_basic/obstacle_detector.h]: This todo needs to either be done, or more detailed

File: src/move_basic.cpp [hunk 4]
```diff
@@ -41,6 +41,7 @@
 #include <geometry_msgs/Vector3.h>
 #include <nav_msgs/Path.h>
 #include <std_msgs/Float32.h>
+#include <std_msgs/UInt8.h>
```
Comment (rohbotics) [inline src/move_basic.cpp]: Should be custom msg type with 'enum' in it. 

Something like
```
uint8 DRIVE_STRAIGHT=0
uint8 FOLLOW_LEFT=1
uint8 FOLLOW_RIGHT=2

uint8 state
```

File: src/move_basic.cpp [hunk 5]
```diff
@@ -395,6 +461,12 @@ void MoveBasic::executeAction(const move_base_msgs::MoveBaseGoalConstPtr& msg)
         }
     }
 
+    // XXX bogus rotation to test linear recovery
```
Comment (rohbotics) [inline src/move_basic.cpp:507]: This is icky, but we can keep it for now

File: src/move_basic.cpp [hunk 6]
```diff
@@ -568,14 +648,91 @@ bool MoveBasic::moveLinear(const tf2::Transform& goalInOdom)
...
+        tf2::Transform initialBaseToCurrent = poseOdom * poseOdomInitial;
+        double cx, cy, cyaw;
+        getPose(initialBaseToCurrent, cx, cy, cyaw);
+
         double distRemaining = remaining.x();
         double distTravelled = std::abs(requestedDistance) - std::abs(distRemaining);
 
+        double velMult = 1.0;
+
+        if (followMode == FOLLOW_LEFT) {
```
Comment (rohbotics) [inline src/move_basic.cpp:745]: Switch statement here? (Doesn't need branches)

File: src/move_basic.cpp [hunk 7]
```diff
@@ -637,16 +814,14 @@ bool MoveBasic::moveLinear(const tf2::Transform& goalInOdom)
             velocity = 0;
         }
 
-        if (std::abs(distTravelled) > std::abs(requestedDistance) - linearTolerance) {
+        // TODO: is this the best test of completeness?
```
Comment (rohbotics) [inline src/move_basic.cpp:920]: No, this is not a great test of completeness. I think the robot should still aim for the x position, not x-tolerance. This should be added as an issue

**Review by rohbotics — CHANGES_REQUESTED** [review]

File: src/map.cpp (hunk 1 above)
Comment (jim-v) [reply src/map.cpp]: Yes, removed.

File: include/move_basic/obstacle_detector.h (hunk 3 above)
Comment (jim-v) [reply include/move_basic/obstacle_detector.h]: Removed the comment.

File: src/move_basic.cpp (hunk 6 above)
Comment (jim-v) [reply src/move_basic.cpp:745]: I can see advantages and disadvantages of both.  `if ... else if` gives a new scope and there is no danger of falling through to the wrong block if a `break` statement is missing.

File: README.md (hunk 2 above)
Comment (jim-v) [reply README.md:29]: Done

**Comment by jim-v** [comment]
@rohbotics, this is ready to be reviewed again when you have time.

**Comment by rohbotics** [comment]
Hey Jim, let me know when you want me to re-review this.

**Comment by jim-v** [comment]
> Hey Jim, let me know when you want me to re-review this.

I'm going to spend some time at the weekend seeing if I can make it work better, but it would be good if you take a look at it before that (assuming you have time), to see if you can spot any potential issues.

**Comment by rohbotics** [comment]
I took a look at it, nothing really stands out as a problem.

There is a lot of close to repetitive code, especially in obstacle detector, with subtle differences. 
If it is easy to parametrize the differences, instead of mostly repeating, that would be nice, but not needed.

**Review by rohbotics — APPROVED** [review]
I think we can merge this. Nothing seems obviously wrong, and if we find some quality problems later on we can fix them.

**Pull request merged**